// testcontainers/dist/modules/arangodb/arangodb-container.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StartedArangoContainer = exports.ArangoDBContainer = void 0;
const __1 = require("../..");
const uuid_1 = require("../../uuid");
const abstract_started_container_1 = require("../abstract-started-container");
class ArangoDBContainer extends __1.GenericContainer {
    constructor(image = "arangodb", tag = "latest", password = new uuid_1.RandomUuid().nextUuid()) {
        super(image, tag);
        this.password = password;
        this.defaultPort = 8529;
        this.defaultUsername = "root";
        this.withExposedPorts(this.defaultPort).withWaitStrategy(__1.Wait.forLogMessage("Have fun!"));
    }
    withPassword(password) {
        this.password = password;
        return this;
    }
    preCreate() {
        return __awaiter(this, void 0, void 0, function* () {
            this.withEnv("ARANGO_ROOT_PASSWORD", this.password);
        });
    }
    start() {
        const _super = Object.create(null, {
            start: { get: () => super.start }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return new StartedArangoContainer(yield _super.start.call(this), this.defaultPort, this.defaultUsername, this.password);
        });
    }
}
exports.ArangoDBContainer = ArangoDBContainer;
class StartedArangoContainer extends abstract_started_container_1.AbstractStartedContainer {
    constructor(startedTestContainer, port, username, password) {
        super(startedTestContainer);
        this.port = port;
        this.username = username;
        this.password = password;
        this.host = this.startedTestContainer.getHost();
        this.port = this.startedTestContainer.getMappedPort(port);
    }
    getTcpUrl() {
        return `tcp://${this.host}:${this.port}/`;
    }
    getHttpUrl() {
        return `http://${this.host}:${this.port}/`;
    }
    getPassword() {
        return this.password;
    }
    getUsername() {
        return this.username;
    }
}
exports.StartedArangoContainer = StartedArangoContainer;
