// @polkadot/react-qr/DisplayAddress.cjs
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.QrDisplayAddress = void 0;

var _react = _interopRequireWildcard(require("react"));

var _Display = require("./Display.cjs");

var _util = require("./util.cjs");

var _jsxRuntime = require("react/jsx-runtime");

// Copyright 2017-2021 @polkadot/react-qr authors & contributors
// SPDX-License-Identifier: Apache-2.0
function DisplayAddress({
  address,
  className,
  genesisHash,
  size,
  style
}) {
  const data = (0, _react.useMemo)(() => (0, _util.createAddressPayload)(address, genesisHash), [address, genesisHash]);

  if (!data) {
    return null;
  }

  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Display.QrDisplay, {
    className: className,
    size: size,
    skipEncoding: true,
    style: style,
    value: data
  });
}

const QrDisplayAddress = /*#__PURE__*/_react.default.memo(DisplayAddress);

exports.QrDisplayAddress = QrDisplayAddress;
