// @polkadot/react-qr/ScanAddress.js
// Copyright 2017-2021 @polkadot/react-qr authors & contributors
// SPDX-License-Identifier: Apache-2.0
import React, { useCallback } from 'react';
import { assert } from '@polkadot/util';
import { decodeAddress } from '@polkadot/util-crypto';
import { ADDRESS_PREFIX, SEED_PREFIX } from "./constants.js";
import { QrScan } from "./Scan.js";
import { jsx as _jsx } from "react/jsx-runtime";

function ScanAddress({
  className,
  onError,
  onScan,
  size,
  style
}) {
  const _onScan = useCallback(data => {
    if (data) {
      try {
        const [prefix, content, genesisHash, ...name] = data.split(':');
        const isValidPrefix = prefix === ADDRESS_PREFIX || prefix === SEED_PREFIX;
        assert(isValidPrefix, `Invalid prefix received, expected '${ADDRESS_PREFIX}/${SEED_PREFIX}' , found '${prefix}'`);
        const isAddress = prefix === ADDRESS_PREFIX;

        if (isAddress) {
          decodeAddress(content);
        }

        onScan({
          content,
          genesisHash,
          isAddress,
          name: name !== null && name !== void 0 && name.length ? name.join(':') : undefined
        });
      } catch (error) {
        console.error('@polkadot/react-qr:QrScanAddress', error.message, data);
      }
    }
  }, [onScan]);

  return /*#__PURE__*/_jsx(QrScan, {
    className: className,
    onError: onError,
    onScan: _onScan,
    size: size,
    style: style
  });
}

export const QrScanAddress = /*#__PURE__*/React.memo(ScanAddress);
