// testcontainers/dist/container.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DockerodeContainer = void 0;
const logger_1 = require("./logger");
class DockerodeContainer {
    constructor(container) {
        this.container = container;
    }
    getId() {
        return this.container.id;
    }
    start() {
        return this.container.start();
    }
    stop(options) {
        return this.container
            .stop({
            t: options.timeout / 1000,
        })
            .catch((error) => {
            /* 304 container already stopped */
            if (error.statusCode === 304) {
                logger_1.log.info(`Container has already been stopped: ${this.getId()}`);
            }
            else {
                throw error;
            }
        });
    }
    remove(options) {
        return this.container.remove({
            v: options.removeVolumes,
        });
    }
    exec(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return new DockerodeExec(yield this.container.exec({
                Cmd: options.cmd,
                AttachStdout: options.attachStdout,
                AttachStderr: options.attachStderr,
            }));
        });
    }
    logs() {
        return __awaiter(this, void 0, void 0, function* () {
            const options = {
                follow: true,
                stdout: true,
                stderr: true,
            };
            return (yield this.container.logs(options)).setEncoding("utf-8");
        });
    }
    inspect() {
        return __awaiter(this, void 0, void 0, function* () {
            const inspectResult = yield this.container.inspect();
            return {
                hostPorts: this.getHostPorts(inspectResult),
                internalPorts: this.getInternalPorts(inspectResult),
                name: this.getName(inspectResult),
                healthCheckStatus: this.getHealthCheckStatus(inspectResult),
                networkSettings: this.getNetworkSettings(inspectResult),
            };
        });
    }
    getName(inspectInfo) {
        return inspectInfo.Name;
    }
    getInternalPorts(inspectInfo) {
        return Object.keys(inspectInfo.NetworkSettings.Ports).map((port) => Number(port.split("/")[0]));
    }
    getHostPorts(inspectInfo) {
        return Object.values(inspectInfo.NetworkSettings.Ports)
            .filter((portsArray) => portsArray !== null)
            .map((portsArray) => Number(portsArray[0].HostPort));
    }
    getHealthCheckStatus(inspectResult) {
        const health = inspectResult.State.Health;
        if (health === undefined) {
            return "none";
        }
        else {
            return health.Status;
        }
    }
    getNetworkSettings(inspectResult) {
        return Object.entries(inspectResult.NetworkSettings.Networks)
            .map(([networkName, network]) => ({
            [networkName]: {
                networkId: network.NetworkID,
                ipAddress: network.IPAddress,
            },
        }))
            .reduce((prev, next) => (Object.assign(Object.assign({}, prev), next)), {});
    }
}
exports.DockerodeContainer = DockerodeContainer;
class DockerodeExec {
    constructor(exec) {
        this.exec = exec;
    }
    start() {
        return new Promise((resolve, reject) => {
            const options = { Detach: false, Tty: true, stream: true, stdin: true, stdout: true, stderr: true };
            this.exec.start(options, (err, stream) => {
                if (err) {
                    return reject(err);
                }
                else if (!stream) {
                    return reject(new Error("Unexpected error occurred, stream is undefined"));
                }
                return resolve(stream);
            });
        });
    }
    inspect() {
        return __awaiter(this, void 0, void 0, function* () {
            const inspectResult = yield this.exec.inspect();
            return {
                exitCode: inspectResult.ExitCode === null ? -1 : inspectResult.ExitCode,
                running: inspectResult.Running,
                entrypoint: inspectResult.ProcessConfig.entrypoint,
                arguments: inspectResult.ProcessConfig.arguments,
            };
        });
    }
}
