// testcontainers/dist/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.containerLog = exports.log = exports.FakeLogger = void 0;
const debug_1 = __importDefault(require("debug"));
class DebugLogger {
    constructor(namespace) {
        this.logger = debug_1.default(namespace);
    }
    trace(message) {
        this.logger(`TRACE ${message}`);
    }
    debug(message) {
        this.logger(`DEBUG ${message}`);
    }
    info(message) {
        this.logger(`INFO ${message}`);
    }
    warn(message) {
        this.logger(`WARN ${message}`);
    }
    error(message) {
        this.logger(`ERROR ${message}`);
    }
}
class FakeLogger {
    constructor() {
        this.traceLogs = [];
        this.debugLogs = [];
        this.infoLogs = [];
        this.warnLogs = [];
        this.errorLogs = [];
    }
    trace(message) {
        this.traceLogs.push(message);
    }
    debug(message) {
        this.debugLogs.push(message);
    }
    info(message) {
        this.infoLogs.push(message);
    }
    warn(message) {
        this.warnLogs.push(message);
    }
    error(message) {
        this.errorLogs.push(message);
    }
}
exports.FakeLogger = FakeLogger;
exports.log = new DebugLogger("testcontainers");
exports.containerLog = new DebugLogger("testcontainers:containers");
