// testcontainers/dist/repo-tag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RepoTag = void 0;
const reaper_1 = require("./reaper");
const port_forwarder_1 = require("./port-forwarder");
class RepoTag {
    constructor(image, tag) {
        this.image = image;
        this.tag = tag;
        this.HELPER_CONTAINERS = new Set([reaper_1.ReaperInstance.IMAGE_NAME, port_forwarder_1.PortForwarderInstance.IMAGE_NAME]);
    }
    equals(repoTag) {
        return this.image === repoTag.image && this.tag === repoTag.tag;
    }
    toString() {
        return `${this.image}:${this.tag}`;
    }
    isReaper() {
        return this.image === reaper_1.ReaperInstance.IMAGE_NAME;
    }
    isHelperContainer() {
        return this.HELPER_CONTAINERS.has(this.image);
    }
    static fromString(string) {
        const parts = string.split("/");
        if (parts.length === 1) {
            return this.fromImage(parts[0]);
        }
        else {
            return this.fromImage(parts[1]);
        }
    }
    static fromImage(image) {
        const [imageName, tag] = image.split(":");
        return new RepoTag(imageName, tag);
    }
}
exports.RepoTag = RepoTag;
