// @zeitgeistpm/type-defs/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.index = exports.swaps = exports.shares = exports.predictionMarkets = exports.orderbook = void 0;
var orderbook_1 = require("./orderbook");
Object.defineProperty(exports, "orderbook", { enumerable: true, get: function () { return __importDefault(orderbook_1).default; } });
var predictionMarkets_1 = require("./predictionMarkets");
Object.defineProperty(exports, "predictionMarkets", { enumerable: true, get: function () { return __importDefault(predictionMarkets_1).default; } });
var shares_1 = require("./shares");
Object.defineProperty(exports, "shares", { enumerable: true, get: function () { return __importDefault(shares_1).default; } });
var swaps_1 = require("./swaps");
Object.defineProperty(exports, "swaps", { enumerable: true, get: function () { return __importDefault(swaps_1).default; } });
exports.index = {
    rpc: {},
    types: {
        Address: "AccountId",
        LookupSource: "AccountId",
        RefCount: "u32",
    },
};
