// testcontainers/dist/modules/abstract-started-container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractStartedContainer = void 0;
class AbstractStartedContainer {
    constructor(startedTestContainer) {
        this.startedTestContainer = startedTestContainer;
    }
    stop(options) {
        return this.startedTestContainer.stop(options);
    }
    getHost() {
        return this.startedTestContainer.getHost();
    }
    getMappedPort(port) {
        return this.startedTestContainer.getMappedPort(port);
    }
    getName() {
        return this.startedTestContainer.getName();
    }
    getId() {
        return this.startedTestContainer.getId();
    }
    getNetworkNames() {
        return this.startedTestContainer.getNetworkNames();
    }
    getNetworkId(networkName) {
        return this.startedTestContainer.getNetworkId(networkName);
    }
    getIpAddress(networkName) {
        return this.startedTestContainer.getIpAddress(networkName);
    }
    exec(command) {
        return this.startedTestContainer.exec(command);
    }
    logs() {
        return this.startedTestContainer.logs();
    }
}
exports.AbstractStartedContainer = AbstractStartedContainer;
